src/treeview_switches.ml

```diff
@@ -0,0 +1,198 @@
+        ; dark = `String (extension_path ^ "/assets/dependency-dark.svg")
+        }
+
+  let collapsible_state = function
+    | Dependency _ -> Vscode.TreeItemCollapsibleState.None
+    | Switch _ -> Vscode.TreeItemCollapsibleState.Collapsed
+end
+
+let make_item ~extension_path ~opam dependency =
+  let open Promise.Syntax in
+  let icon = `LightDark (Dependency.icon ~extension_path dependency) in
+  let collapsibleState = Dependency.collapsible_state dependency in
+  let label =
+    Vscode.TreeItemLabel.create ~label:(Dependency.label dependency) ()
+  in
+  let item = Vscode.TreeItem.make ~label ~collapsibleState () in
+  TreeItem.set_id item (Dependency.to_string dependency);
+  TreeItem.set_iconPath item icon;
+  TreeItem.set_contextValue item (Dependency.context_value dependency);
+  let+ _ =
+    Promise.Option.iter
+      (fun desc -> TreeItem.set_description item (`String desc))
+      (Dependency.description ~opam dependency)
+  in
+  Option.iter (Dependency.tooltip dependency) ~f:(fun desc ->
+      TreeItem.set_tooltip item (`String desc));
+  item
+
+let get_dependency_dependencies _dependency = Promise.return (Some [])
+
+let get_switch_dependencies opam switch =
+  let open Promise.Syntax in
+  let* packages = Opam.get_switch_packages opam switch in
+  match packages with
+  | Ok packages ->
+    let names =
+      packages
+      |> List.map ~f:(fun n ->
+             Dependency.Dependency (n, Some (Dependency.Switch switch)))
+    in
+    Promise.return (Some names)
+  | Error err ->
+    let* _ =
+      Vscode.Window.showInformationMessage
+        ~message:
+          (Printf.sprintf
+             "An error occured while reading the switch dependencies: %s" err)
+        ()
+    in
+    Promise.return None
+
+let get_dependencies ~opam ~extension_path element =
+  let open Promise.Syntax in
+  let dependency =
+    match Vscode.TreeItem.id element with
+    | None ->
+      (* This should never happen, the elements are always created with an ID. *)
+      assert false
+    | Some dep -> Dependency.of_string dep
+  in
+  let dependencies_opt_to_tree_items deps_opt =
+    let* deps_opt = deps_opt in
+    match deps_opt with
+    | None -> Promise.return None
+    | Some deps ->
+      let* items =
+        List.map deps ~f:(make_item ~opam ~extension_path) |> Promise.all_list
+      in
+      Promise.return (Some items)
+  in
+  match dependency with
+  | Dependency.Switch switch ->
+    let deps = get_switch_dependencies opam switch in
+    dependencies_opt_to_tree_items deps
+  | Dependency.Dependency dependency ->
+    let deps = get_dependency_dependencies dependency in
+    dependencies_opt_to_tree_items deps
+
+let register extension =
+  let open Promise.Syntax in
+  let extension_path = Vscode.ExtensionContext.extensionPath extension in
+  let+ opam = Opam.make () in
+  let getChildren ~element :
+      [ `Promise of TreeItem.t list option Promise.t
+      | `Value of TreeItem.t list option
+      ] =
```

Comment:
This type annotation can be `TreeItem.t list Vscode.ProviderResult.t`

Reply:
Thanks @mnxn, just added it to debug something, I'll remove it instead